Fatras/include/ActsFatras/Kernel/detail/Interactor.hpp

```diff
@@ -96,14 +100,26 @@ struct Interactor {
     // decay check. needs to happen at every step, not just on surfaces.
     // TODO limit the stepsize when close to the lifetime limit to avoid
     //   overstepping and decaying the particle systematically too late
-    if (result.properTimeLimit < result.particle.properTime()) {
+    if ((result.properTimeLimit - result.particle.properTime() < result.properTimeLimit * properTimeTolerance) &&
+        result.isAlive) {
       auto descendants = decay.run(generator, result.particle);
       for (auto &&descendant : descendants) {
         result.generatedParticles.emplace_back(std::move(descendant));
       }
       result.isAlive = false;
       return;
     }
+    
+    // Regulate the step size 
+    if(std::isfinite(result.properTimeLimit)) {
+		//    beta² = p²/E²
+		//    gamma = 1 / sqrt(1 - beta²) = sqrt(m² + p²) / m = E / m
+	    //     time = proper-time * gamma
+	    // ds = beta * dt = (p/E) dt (E/m) = (p/m) proper-time
+		const auto properTimeDiff = result.properTimeLimit - result.particle.properTime();
+		const auto stepSize = properTimeDiff * result.particle.absoluteMomentum() / result.particle.mass();
```

Comment:
Can we have zero mass here?

Reply:
This somewhat belongs to your previous comment: If you check for a finite lifetime then photons are excluded. So for a proper setup I don't see how you could come up with unstable particles with zero mass.

Reply:
Ok, makes sense. Could you add a comment to that affect?

Reply:
Comment added